pkg/engine/task/task_pipe.go

```diff
@@ -0,0 +1,427 @@
+	if len(ctr.VolumeMounts) != 1 {
+		return errors.New("pipe container should have exactly one volume mount")
+	}
+
+	if ctr.ReadinessProbe != nil {
+		return errors.New("pipe container does not support readiness probes")
+	}
+
+	// check if all referenced pipe files are children of the container mountPath
+	mountPath := ctr.VolumeMounts[0].MountPath
+	for _, f := range ff {
+		if !isRelative(mountPath, f.File) {
+			return fmt.Errorf("pipe file %s should be a child of %s mount path", f.File, mountPath)
+		}
+
+		fileName := path.Base(f.File)
+		// k8s uses the same validation for Secret/ConfigMap data keys: a valid key must consist of alphanumeric
+		// characters, '-', '_' or '.' (e.g. 'key.name',  or 'KEY_NAME',  or 'key-name', regex used for
+		// validation is '[-._a-zA-Z0-9]+')
+		if !pipeFileRe.MatchString(fileName) {
+			return fmt.Errorf("pipe file name %s should only contain alphanumeric characters, '.', '_' and '-'", fileName)
+		}
+	}
+	return nil
+}
+
+func pipePod(ctr *corev1.Container, name string) (string, error) {
+	volumeName := ctr.VolumeMounts[0].Name
+	mountPath := ctr.VolumeMounts[0].MountPath
+
+	pod := corev1.Pod{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Pod",
+			APIVersion: "v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
+		Spec: corev1.PodSpec{
+			Volumes: []corev1.Volume{
+				{
+					Name:         volumeName,
+					VolumeSource: corev1.VolumeSource{EmptyDir: nil},
+				},
+			},
+			InitContainers: []corev1.Container{*ctr},
+			Containers: []corev1.Container{
+				{
+					Name:    pipePodContainerName,
+					Image:   "busybox",
+					Command: []string{"/bin/sh", "-c"},
+					Args:    []string{"sleep infinity"},
+					VolumeMounts: []corev1.VolumeMount{
+						{
+							Name:      volumeName,
+							MountPath: mountPath,
+						},
+					},
+				},
+			},
+			RestartPolicy: corev1.RestartPolicyOnFailure,
+		},
+	}
+
+	b, err := yaml.Marshal(pod)
+	if err != nil {
+		return "", fmt.Errorf("failed to create pipe task pod: %v", err)
+	}
+
+	return string(b), nil
+}
+
+func copyFiles(fs afero.Fs, ff []PipeFile, pod *corev1.Pod, ctx Context) error {
+	restCfg, err := config.GetConfig()
+	if err != nil {
+		return fatalExecutionError(fmt.Errorf("failed to fetch cluster REST config: %v", err), pipeTaskError, ctx.Meta)
+	}
+
+	var g errgroup.Group
+
+	for _, f := range ff {
+		f := f
+		log.Printf("PipeTask: %s/%s copying pipe file %s", ctx.Meta.InstanceNamespace, ctx.Meta.InstanceName, f.File)
+		g.Go(func() error {
+			return copyFile(fs, f, pod, restCfg)
+		})
+	}
+
+	err = g.Wait()
+	return err
+}
+
+func copyFile(fs afero.Fs, pf PipeFile, pod *corev1.Pod, restCfg *rest.Config) error {
+	reader, stdout := io.Pipe()
+
+	defer reader.Close()
+	defer stdout.Close()
+
+	pe := PodExec{
+		RestCfg:       restCfg,
+		PodName:       pod.Name,
+		PodNamespace:  pod.Namespace,
+		ContainerName: pipePodContainerName,
+		Args:          []string{"tar", "cf", "-", pf.File},
```

Comment:
I wonder if it would be worth doing the file size validation somewhere here already. If we do it in the pipeFiles, we may already have copied gigabytes of stuff from the initContainer to a local memFs...

Reply:
I was thinking about that too. We would need to run a command determining the file size, parsing its output and then run another one to download it. I'll get back to it once I wanted to get back to it once I'm done with the rest of the PR.

Reply:
And done!